[Production Release- 1.2.0] Tech Space Consent Flow

Visions-David opened the pull request:
- Protecting all new routes with a Header : "x-visionstrust-consent-key"
- Adding protected route to get a userIdentifier by email and participant self description and return if exists and if is linked to a user
- Adding protected route to get a user by email and return user
- Add protected route to inject userIdentifier in User

Visions-David committed:
feat : consent key check middleware creation

Visions-David committed:
chore : .env sample update to implement visionstrust consent key

Visions-David committed:
feat : user router update to take in account new user functionalities

Visions-David committed:
feat : user controller update to take in account new user functionalities

Visions-David committed:
fix : error handling fix

Visions-David committed:
refactor : body params refactoring on user identifier injection method

Visions-Nicolas commented on src/controllers/usersController.ts:
```diff
@@ -286,3 +288,87 @@ export const registerUserIdentifiers = async (
+
+  const participant = await Participant.findOne({
+    selfDescriptionURL: selfDescription,
+    email: email,
```
the email is not needed, it's the email of the userIdentifier not of the participant

Visions-Nicolas commented on src/routes/users.ts:
```diff
@@ -32,6 +36,12 @@ r.delete("/logout", async (req, res, next) => {
 
+r.get("/useridentifierbyemail", consentKeyCheck, getUserIdentifierByEmail);
+
+r.get("/userbyemail", consentKeyCheck, getUserByEmail);
```
Here the routes must be rename to be more comprehensive and the HTTP verb updated (it's not common for a GET request to have a body):
- `r.get("/useridentifierbyemail", consentKeyCheck, getUserIdentifierByEmail); `→` r.post("/identifier/search", consentKeyCheck, getUserIdentifierByEmail);`
-` r.get("/userbyemail", consentKeyCheck, getUserByEmail)`; → `r.post("/search", consentKeyCheck, getUserByEmail);`

Visions-Nicolas commented on src/routes/users.ts:
```diff
@@ -32,6 +36,12 @@ r.delete("/logout", async (req, res, next) => {
+r.get("/useridentifierbyemail", consentKeyCheck, getUserIdentifierByEmail);
+
+r.get("/userbyemail", consentKeyCheck, getUserByEmail);
+
+r.post("/useridentifier", consentKeyCheck, injectUserIdentifier);
+
```
rename to `/identifier`

Visions-Nicolas commented on src/controllers/usersController.ts:
```diff
@@ -286,3 +288,87 @@ export const registerUserIdentifiers = async (
+  }
+
+  return res.status(200).json(results);
+};
+
+export const getUserByEmail = async (req: Request, res: Response) => {
+  const { email } = req.body;
+  const user = await User.findOne({
+    email: email,
+  }).lean();
+
+  if (!user) {
+    return res.status(404).json({ message: "User not found" });
+  } else {
+    return res.status(200).json(user);
+  }
+};
```
We need to hide the `oauth` and `password` field for security reason

Visions-David committed:
fix : body modification and user getter refactoring to avoid fields

Visions-David committed:
fix : users router names and method fix

Visions-Nicolas commented on src/controllers/usersController.ts:
```diff
@@ -296,7 +296,7 @@ export const registerUserIdentifiers = async (
  * @returns
  */
 export const getUserIdentifierByEmail = async (req: Request, res: Response) => {
-  const { email, selfDescription } = req.body;
```
email is still needed to find the userIdentifier

Visions-Nicolas commented on src/controllers/usersController.ts:
```diff
@@ -305,15 +305,13 @@ export const getUserIdentifierByEmail = async (req: Request, res: Response) => {
 
   const userIdentifier = await UserIdentifier.findOne({
     attachedParticipant: new mongoose.Types.ObjectId(participant._id),
-    email: email,
```
undo

Visions-Nicolas commented on src/controllers/usersController.ts:
```diff
@@ -322,7 +320,7 @@ export const getUserIdentifierByEmail = async (req: Request, res: Response) => {
     results.userIdentifier = userIdentifier._id;
     const user = await User.findOne({
-      email: email,
+      email: userIdentifier.email,
```
undo

Visions-David committed:
fix : rollback email use for user identifier getter

Visions-Nicolas committed:
fix: moved the server initialization in test to be in describe, now the tests passed
fix: injectUserIdentifier to respond with the user with updated identifiers and replace map with a for loop
feat: added tests for new routes

Visions-Nicolas committed:
feat: added morganLogs

Visions-Nicolas committed:
refactor: delete console.log in tests

Visions-David committed:
refactor : user controller error handling on get user identifier by email method refactoring

Visions-David committed:
fix : variable typo fix

Visions-Nicolas committed:
refactor: update getUserIdentifierByEmail
test: added test for getUserIdentifierByEmail

Visions-David committed:
fix : user identifier by email getter user existence check fix

Visions-Nicolas committed:
feat: added participant Id on getUserIdentifierByEmail

Visions-Nicolas committed:
fix: avoid generation contract URI when generating privacy notice

Visions-Nicolas committed:
feat: avoid pushing the same identifiers multiples times in injectUserIdentifier

Visions-Nicolas committed:
core: up version to 1.2.0
